### PR TITLE
applied HF image transformer block

### LINKED ISSUE #30087: `test_llama_image_block.py`
replace llama_models import with hf import is the end goal.

https://github.com/tenstorrent/tt-metal/blob/bf338f0844e710e58de614b919f4e959b6f133a1/models/tt_transformers/tests/multimodal/test_llama_image_block.py#L57-L59

### PR DESCRIPTION
**Pull request by epam-ioannis-alexiou**
### Ticket
Close https://github.com/tenstorrent/tt-metal/issues/30087

### Problem description
vision branch has several repeated image transformer blocks, this PR applies the equivalent of meta lib using HF and compares tensor similarity with TT implementation for image and global image transformer blocks.

### What's changed
The unmapped HF weights are used for the HF model whereas TT model uses remapped HF weights. A function to load and extract the revelavant weights is used so the weights can be allocated to the reference layer. Also HF build the attention maps internally in this example the attention maps are explicitly defined in the script and passed to the attention blocks.

Note:
HF internally may use different attention implementations, by default eager mode is always active. This mode uses an attention implementation that if combined with making a continuous block of memory using .contiguous() and nn.Linear layer generates minor numerical deviations in the generated tensor. Thus when testing Llama model attention its implementation is switched to the following
"config.vision_config._attn_implementation = "sdpa"" to use torch.nn.functional.scaled_dot_product_attention as in meta lib (this diminishes any numerical fluctuations).

### Checklist
- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/20367495677) CI passes  for `test_llama_image_block.py` with relevant CI tasks `t3k llama3.2-90b-vision tests`, `t3k n300 mesh llama3.2-11b-vision tests`, `t3k llama3.2-11b-vision tests`

### CODE REVIEW

File: models/tt_transformers/tests/multimodal/test_llama_image_block.py [hunk 1]
```diff
@@ -3,20 +3,53 @@
...
+    tensor = F.pad(tensor, padding, mode="constant", value=0)
+    slice_index = -num_padding_patches if num_padding_patches > 0 else 0
+    return tensor, slice_index
+
+
+def contract_num_tokens_from_mult8(tensor, slice_index):
+    if slice_index == 0:
+        return tensor
+    return tensor[:, :, :slice_index, :]
+
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_block.py]: All three functions could be refactored into `models/tt_transformers/tests/multimodal/utils.py`.

File: models/tt_transformers/tests/multimodal/test_llama_image_block.py [hunk 2]
```diff
@@ -45,18 +78,21 @@ def test_block_inference(batch, num_chunks, mesh_device, gated, reset_seeds, ens
     # Ref model needs partial state dict, but our models use full state dict keys as cached weight names
     if gated:
         first_layer_prefix = "vision_model.vision_encoder.global_transformer.resblocks.0."
+        hf_layer_prefix = "model.vision_model.global_transformer.layers.0."
     else:
         first_layer_prefix = "vision_model.vision_encoder.transformer.resblocks.31."
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_block.py:56]: are the `first_layer_prefix` instances used in the code anymore? if not, they should be deleted.

File: models/tt_transformers/tests/multimodal/test_llama_image_block.py (hunk 1 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_block.py]: done in [1991055](https://github.com/tenstorrent/tt-metal/pull/34487/commits/1991055c18e0878deea2783a346a50509982998c)

File: models/tt_transformers/tests/multimodal/test_llama_image_block.py (hunk 2 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_block.py:56]: `first_layer_prefix ` is used by tt_model at line 76 while `hf_layer_prefix` is for the reference model. Boht use diffferent naming conventions but practically use the same weights.

File: models/tt_transformers/tests/multimodal/test_llama_image_block.py [hunk 3]
```diff
@@ -76,9 +85,9 @@ def test_block_inference(batch, num_chunks, mesh_device, gated, reset_seeds, ens
     pt_block_input = (torch.rand(batch, num_chunks, ntok, dim) * 2) - 1
     tt_attention_input = pt_block_input.clone()
     # Do PT padding
-    pt_block_input, npad = encoder_utils.expand_num_tokens_to_mult8(pt_block_input)
+    pt_block_input, slice_index = expand_num_tokens_to_mult8(pt_block_input)
```
Comment (maksim-tsishkouski-epam) [inline models/tt_transformers/tests/multimodal/test_llama_image_block.py:88]: btw, does HF need to expand to mult8? or is it only llama_models functionality?

Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_block.py:88]: yes HF is doing this also for the whole model here [padding logic](https://github.com/huggingface/transformers/blob/c67ec2c4c14477faf7b258eba532d2673ddad4a7/src/transformers/models/mllama/modeling_mllama.py#L1113). I transferred the HF logic using the same function names as meta lib while the implementation difffers a bit.

**Review by maksim-tsishkouski-epam — APPROVED** [review]

File: models/tt_transformers/tests/multimodal/utils.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+import os
+
+import torch.nn.functional as F
+
+
+def load_partial_weights(auto_model, weights_path, layer_prefix):
```
Comment (maksim-tsishkouski-epam) [inline models/tt_transformers/tests/multimodal/utils.py:9]: maybe modify other files as well to use this function

Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/utils.py:9]: https://github.com/search?q=repo%3Atenstorrent%2Ftt-metal%20load_partial_weights&type=code shows that there are currently four inlined `load_partial_weights`.

Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/utils.py:9]: done in [f32721](https://github.com/tenstorrent/tt-metal/pull/34487/commits/7f3272187f1e33ffb12b40d9f6d64c9604ee3217)

**Review by gwangTT — APPROVED** [review]
Thanks Ioannis, for reducing the technical debt! The code looks much cleaner now!
